### PR TITLE
feat(graindoc): Refactored Graindoc parser

### LINKED ISSUE #971: Graindoc: Support multi-line examples
Discussed in discord with @spotandjake

This was originally an issue because we processed each line with the same regexp, but we now have a double-pass system where each attribute parses itself. I'm wondering if we can expand the first (simpler) attribute regexp to parse multiple lines as an attribute.

### LINKED ISSUE #1653: Graindoc: Use proper lexer?
Should graindoc use a proper lexer, instead of 2 passes of regexps?

### PR DESCRIPTION
**Pull request by ospencer**
This PR implements a new parser for Graindoc, providing more accurate errors and expanding its capabilities. Namely, attribute descriptions now all support multiple lines, including `@example`!

Either syntax works fine:
```
/**
 * @example let foo = "foo"
 * print(foo)
 */
```

```
/**
 * @example
 * let foo = "foo"
 * print(foo)
 */
```

closes #971 
closes #1653

### CODE REVIEW

File: compiler/src/diagnostics/graindoc_lexer.re [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  } else {
+    semver;
+  };
+
+// HORIZONTAL TABULATION
+// VERTICAL TABULATION
+// SPACE
+// LEFT-TO-RIGHT MARK
+// RIGHT-TO-LEFT MARK
+let blank = [%sedlex.regexp? 0x09 | 0x0B | 0x20 | 0x200E | 0x200F];
```
Comment (phated) [inline compiler/src/diagnostics/graindoc_lexer.re:42]: Are some of these defined in `Grain_parsing`? Do we need to redefine them or can we share them?

File: compiler/src/diagnostics/graindoc_lexer.re [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+and default = (state, lexbuf) => {
+  switch%sedlex (lexbuf) {
+  | asterisk => default(state, lexbuf)
+  | "@param" =>
+    state.lexer_mode = Param;
+    PARAM;
+  | "@returns" =>
+    state.lexer_mode = GenericAttribute;
```
Comment (phated) [inline compiler/src/diagnostics/graindoc_lexer.re:100]: Should this be called like `UnlabeledAttribute` or `FreeTextAttribute`?

File: compiler/src/diagnostics/comment_attributes.re [hunk 3]
```diff
@@ -0,0 +1,37 @@
+exception InvalidAttribute(string);
```
Comment (phated) [inline compiler/src/diagnostics/comment_attributes.re:1]: Why did you split this into a different file?

File: compiler/src/diagnostics/graindoc_parser.mly [hunk 4]
```diff
@@ -0,0 +1,48 @@
+%token PARAM SECTION SINCE HISTORY THROWS RETURNS EXAMPLE DEPRECATED COLON EOL EOF
+%token <string> TEXT IDENT SEMVER CONSTRUCTOR
+
+%right EOL
+
+%start <(Comment_attributes.parsed_graindoc)> graindoc
```
Comment (phated) [inline compiler/src/diagnostics/graindoc_parser.mly:6]: Can this just be `(option(string), list(t))`
```suggestion
%start <((option(string), Comments.Attribute.attributes))> graindoc
```

**Review by phated — CHANGES_REQUESTED** [review]
This is so awesome! Thanks for looking into it 🙏 I had a few questions

File: compiler/src/diagnostics/comment_attributes.re (hunk 3 above)
Comment (ospencer) [reply compiler/src/diagnostics/comment_attributes.re:1]: There was a circular dependency, parser -> comments -> parser 😢

File: compiler/src/diagnostics/graindoc_lexer.re (hunk 1 above)
Comment (ospencer) [reply compiler/src/diagnostics/graindoc_lexer.re:42]: I tried to share them, but it didn't seem possible.

File: compiler/src/diagnostics/graindoc_parser.mly (hunk 4 above)
Comment (ospencer) [reply compiler/src/diagnostics/graindoc_parser.mly:6]: I tried that originally, but it's a syntax error. Seems like you have to provide a named type here.

**Review by phated — APPROVED** [review]
LGTM! Thanks 💪

**Pull request merged**